File: Python/Data Structures/Linked List/3_insertion.py
class Node:
    def __init__(self, item):
        self.data = item
        self.next = None


class LinkedList:
    def __init__(self):
        self.head = None

    def createList(self):
        num_elements = int(input("Enter the number of elements in list: "))

        for i in range(num_elements):
            data = int(input(f"Enter the data for node {i + 1}: "))
            newnode = Node(data)

            if self.head is None:
                self.head = newnode
            else:
                temp = self.head
                while temp.next:
                    temp = temp.next
                temp.next = newnode

        self.displayList()

    def displayList(self):
        temp = self.head

        print("Elements in list: [ ", end=" ")
        while temp:
            print(f"{temp.data}", end=" -> ")
            temp = temp.next
        print("NULL ]")

    def insertAtBeginning(self, item: int):
        newnode = Node(item)

        if self.head is None:
            self.head = newnode
            return
        else:
            newnode.next = self.head
            self.head = newnode

        self.displayList()

    def insertAtEnd(self, item: int):
        newnode = Node(item)
        temp = self.head

        while temp.next:
            temp = temp.next
        temp.next = newnode
        self.displayList()

    def insertAtPosition(self, item: int):
        newnode = Node(item)
        count = 0
        i = 0

        current = self.head
        while current:
            current = current.next
            count += 1
        position = int(input("Enter the position: "))

        if position > count or position < 0:
            print("INVALID POSITION")
            return

        temp = self.head
        while i < position - 1:
            temp = temp.next
            i += 1
        newnode.next = temp.next
        temp.next = newnode

        self.displayList()


def main():
    lnkd = LinkedList()
    lnkd.createList()
    lnkd.insertAtBeginning(100)
    lnkd.insertAtEnd(1000)
    lnkd.insertAtPosition(2000)


if __name__ == "__main__":
    main()

File: C/Data_Structures/Linked List/Singly Linked List/3_insertion.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
  int data;
  struct Node *next;
};

struct Node *head;
struct Node *temp;
struct Node *newnode;
int count;

/* Create a linked list */
void createLinkedList() {
  // initialise head null
  head = 0;

  printf("Enter the number of elements in the list: ");
  scanf("%d", &count);

  // create newnode
  for (int i = 0; i < count; i++) {
    newnode = (struct Node *)malloc(sizeof(struct Node));
    printf("Enter the data for %d st node: ", i);
    scanf("%d", &newnode->data);
    newnode->next = 0;

    if (head == 0) {
      head = newnode;
    } else {
      temp = head;
      while (temp->next != 0) {
        temp = temp->next;
      }
      temp->next = newnode;
    }
  }
}

/* Insert node at the beginning */
void insertAtBeginning(int item) {
  newnode = (struct Node *)malloc(sizeof(struct Node));
  newnode->data = item;
  count++;

  if (head == 0) {
    head = newnode;
    /*
     * If there's only one node then
     * no need to further execute
     */
    return;
  } else {
    newnode->next = head; // store the address of first node in the newnode
    head = newnode;       // store address of new node into head
  }
}

/* Insert node at the end */
void insertAtEnd(int item) {
  newnode = (struct Node *)malloc(sizeof(struct Node));
  newnode->data = item;
  count++;

  if (head == 0) {
    head = newnode;
  } else {
    temp = head;
    while (temp->next != 0) {
      temp = temp->next;
    }
    temp->next = newnode;
  }
}

/* Insert node at certain position */
void insertAtPosition(int item) {
  newnode = (struct Node *)malloc(sizeof(struct Node));
  newnode->data = item;
  count++;
  int pos, i = 0;

  printf("Enter the position: ");
  scanf("%d", &pos);

  if (pos > count) {
    printf("\nINVALID POSITION!");
  } else if (pos == 0) {
    newnode->next = head;
    head = newnode;
  } else {
    temp = head;
    while (i < pos - 1) {
      temp = temp->next;
      i++;
    }
    newnode->next = temp->next;
    temp->next = newnode;
  }
}

/* display list */
void displayList() {
  printf("\n[ ");
  temp = head;
  while (temp != 0) {
    printf("%d->", temp->data);
    temp = temp->next;
  }
  printf("NULL ]\n");
  printf("Number of elements in the list: %d\n\n", count);
}

int main(int argc, char *argv[]) {
  createLinkedList();
  insertAtBeginning(5);
  insertAtEnd(60);
  insertAtPosition(100);
  displayList();
  return 0;
}
